### ISSUE TITLE
Different behaviour between typing.Union and | operator

### ISSUE THREAD

**Issue by BrandonNav** [microsoft/pyright#5256]
**Describe the bug**
I was updating code to use the `|` operator over typing.Union
I was surprised that making this change triggered a type error in pyright.

Note: the type error makes sense to me. I understand why the error occured (unknown type, etc).
I just wanted to point out a behaviour difference with something that I thought would be identical.

**To Reproduce**
1. In a clean virtualenv, install:
* django==4.2.2
* django-stubs==4.2.1

2. Create a file with the following code:

```python
from __future__ import annotations

from typing import Union

from django.db import models
from django.db.models import QuerySet


class Contact(models.Model):
    name = models.CharField(max_length=255)


QSLike1 = QuerySet | list[Contact]  # type of QuerySet is QuerySet[Unknown]

def loop1(qs: QSLike1) -> list[tuple[int, Contact]]:
    return [(i, item) for i, item in enumerate(qs)]  # type error here
    #  error: Expression of type "list[tuple[int, _T@QuerySet | Contact]]" cannot be assigned to return  type "list[tuple[int, Contact]]" (reportGeneralTypeIssues)


QSLike2 = Union[QuerySet, list[Contact]]

def loop2(qs: QSLike2) -> list[tuple[int, Contact]]:
    return [(i, item) for i, item in enumerate(qs)]  # no error here
```

3. Run pyright

**Expected behavior**
Both lines should error

**VS Code extension or command-line**
Python: 3.10.12
Pyright: 1.1.3.13
Pylance: prerelease v2023.6.11

**Comment by erictraut** [comment]
Thanks for reporting. I agree this is a bug, but neither lines should generate an error. You're defining a type alias here, and you haven't specified a type argument for the generic class `QuerySet`. That means it should be interpreted as `QuerySet[Unknown]`. If your intent is for `QSLike` to be a generic type alias, you should include a type variable in its definition: `QLike = QuerySet[T] | list[Contact]`.

**Comment by erictraut** [comment]
This will be fixed in the next release. The same bug caused `reportMissingTypeArgument` to not report an error in your use of the generic `QuerySet` type alias without supplying a type argument.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.314, which I just published. It will also be included in this week's Pylance prerelease build.
